### PR TITLE
Allow empty body for json input

### LINKED ISSUE #19171: RESTEasy reactive version presents different behaviour than the non-reactive version
### Describe the bug

I'm trying to make a REST backend application using resteasy, hibernate-validator, oidc and kotlin extensions.

### Expected behavior

The quarkus reactive and non-reactive versions of the resteasy extension have the same behaviour

### Actual behavior

The extensions behave differently:

While testing my backend I found that the evaluation of the constraints present in my model where taking precedence over 
the annotation RolesAllowed which in the non-reactive version of the resteasy extension are being evaluated before.

I also found that the Accept-Language header is not being taken into account in while using the reactive version of the resteasy extension and in the non-reactive one are working ok.

### How to Reproduce?

Reproducer: https://github.com/javiertoja/stackoverflow/tree/main/resteasy-validation

This repository contains two projects (resteasy-validation & resteasy-reactive-validation), both implement the same functionality, a greeting resource. One uses the reactive version of the resteasy extension and the other don't.

Each project contains the same test suite with three test cases which test the following scenarios:
1. A non authorized request to a secured endpoint
2. A bad request response to a request with language acceptation conditions

### Output of `uname -a` or `ver`

Linux nidavellir 5.8.0-63-generic #71~20.04.1-Ubuntu SMP Thu Jul 15 17:46:08 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk 16.0.1 2021-04-20 OpenJDK Runtime Environment AdoptOpenJDK-16.0.1+9 (build 16.0.1+9) OpenJDK 64-Bit Server VM AdoptOpenJDK-16.0.1+9 (build 16.0.1+9, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.1.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by geoand**
Fixes part of #19171

**Edit:**

Closes #19171 

Adding ^ to ensure that the linked issue gets closed when this is merged

### CODE REVIEW

**Review by stuartwdouglas — APPROVED** [review]

File: extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/runtime/serialisers/ServerJacksonMessageBodyReader.java [hunk 1]
```diff
@@ -53,10 +54,22 @@ public Object readFrom(Class<Object> type, Type genericType, MediaType mediaType
...
+            if (isEmptyInputException(e)) {
+                return null;
+            }
+            throw e;
+        }
+    }
+
+    private boolean isEmptyInputException(MismatchedInputException e) {
+        // this isn't great, but Jackson doesn't have a specific exception for empty input...
+        return e.getMessage().startsWith("No content");
```
Comment (FroMage) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/runtime/serialisers/ServerJacksonMessageBodyReader.java:73]: 🤣

Comment (geoand) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/runtime/src/main/java/io/quarkus/resteasy/reactive/jackson/runtime/serialisers/ServerJacksonMessageBodyReader.java:73]: The difficlties of dealing with the real world...

**Comment by geoand** [comment]
I did. But I didn't come to any satisfactory conclusion

**Comment by FroMage** [comment]
Haha. Define satisfactory? The spec (https://quarkus.io/specs/jaxrs/2.1/index.html#standard_entity_providers) says:

> When reading zero-length message entities all pre-packaged MessageBodyReader implementations, except the JAXB one and those for the (boxed) primitive types above, MUST create a corresponding Java object that represents zero-length data. The pre-packaged JAXB and the pre-packaged primitive type MessageBodyReader implementations MUST throw a NoContentException for zero-length message entities.

> When a NoContentException is thrown while reading a server request entity from a MessageBodyReader it MUST be translated by the server runtime into a BadRequestException wrapping the original NoContentException and re-thrown to be processed by any registered exception mappers.

**Comment by FroMage** [comment]
So… I'm going to guess that this is to be able to disambiguate a body with `null` (which is valid JSON) with an empty body?

Now, the question is whether this is valuable. I do wonder how we serialise a `null` value over JSON, now: do we produce `null` or do we throw or send a `NO CONTENT`?

**Comment by FroMage** [comment]
https://quarkus.io/specs/jaxrs/2.1/index.html#resource_method_return says:

> If the return value is not null a 200 status code is used, a null return value results in a 204 status code.

Which means returning a `null` JSON value produces a no-content body, and not a `null` value. So there's no way to produce a JSON `null` value, actually. Funny. But then it's not symmetrical with the fact that we turn an empty body into a `null` value on `POST`. 

Frankly, I think this is not very important. And if the TCK kept passing, it means they didn't even test this part of the spec. So :shrug:

**Comment by geoand** [comment]
> Now, the question is whether this is valuable. I do wonder how we serialise a `null` value over JSON, now: do we produce `null` or do we throw or send a `NO CONTENT`?

The latter

**Comment by geoand** [comment]
> Frankly, I think this is not very important. And if the TCK kept passing, it means they didn't even test this part of the spec. So

Exactly, that's why I didn't care more about it :)

**Pull request merged**